doc/users/next_whats_new/sphinx_plot_preserve.rst

```diff
@@ -0,0 +1,32 @@
+Caching sphinx directive figure outputs
+---------------------------------------
+
+The new ``:outname:`` property for the Sphinx plot directive can
+be used to cache generated images. It is used like:
+
+.. code-block:: rst
+
+    .. plot::
+       :outname: stinkbug_plot
+
+       import matplotlib.pyplot as plt
+       import matplotlib.image as mpimg
+       import numpy as np
+       img = mpimg.imread('_static/stinkbug.png')
+       imgplot = plt.imshow(img)
+
+Without ``:outname:``, the figure generated above would normally be called,
+e.g. :file:`docfile3-4-01.png` or something equally mysterious. With
+``:outname:`` the figure generated will instead be named
+:file:`stinkbug_plot-01.png` or even :file:`stinkbug_plot.png`. This makes it
+easy to understand which output image is which and, more importantly, uniquely
```

Comment:
I am slightly skeptical of the first claim (that users can give things good names) and _very_ skeptical of the second as the name is not random, it is based on the name of the file (and the position + function of the directive in that file) so is structurally unique.  If users can control the names there is a high chance that they are going to collide.

If we want to make things save to cache like this we should give them names based on the hash of the source....